kvserver: pebble panic from WriteSyncNoop

**Describe the problem**

In [a recent CI failure](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_UnitTests_BazelUnitTests/5272407?buildTab=overview&showRootCauses=false&expandBuildChangesSection=true&expandBuildDeploymentsSection=true&expandBuildProblemsSection=true&expandBuildTestsSection=true) a number of TestExecBuild tests failed to return results; however the root cause seems to be a panic from the liveness system:

```
panic: pebble: closed

goroutine 97893 [running]:
github.com/cockroachdb/pebble.(*DB).Apply(0x0, 0x56fb700, 0xc006fa6000)
	github.com/cockroachdb/pebble/external/com_github_cockroachdb_pebble/db.go:717 +0x3c9
github.com/cockroachdb/pebble.(*Batch).Commit(...)
	github.com/cockroachdb/pebble/external/com_github_cockroachdb_pebble/batch.go:932
github.com/cockroachdb/cockroach/pkg/storage.(*pebbleBatch).Commit(0xc01a0959e0, 0x0)
	github.com/cockroachdb/cockroach/pkg/storage/pebble_batch.go:495 +0x3c
github.com/cockroachdb/cockroach/pkg/storage.WriteSyncNoop({0x57cd8f8, 0xc01a0959e0})
	github.com/cockroachdb/cockroach/pkg/storage/engine.go:938 +0xba
github.com/cockroachdb/cockroach/pkg/kv/kvserver/liveness.(*NodeLiveness).updateLiveness.func1()
	github.com/cockroachdb/cockroach/pkg/kv/kvserver/liveness/liveness.go:1274 +0x25
github.com/cockroachdb/cockroach/pkg/util/syncutil/singleflight.(*Group).doCall(0xc0291f7890, 0xc02ef44a50, {0x42b9a51, 0x1}, 0xc033470bd0)
	github.com/cockroachdb/cockroach/pkg/util/syncutil/singleflight/singleflight.go:128 +0x3b
created by github.com/cockroachdb/cockroach/pkg/util/syncutil/singleflight.(*Group).DoChan
	github.com/cockroachdb/cockroach/pkg/util/syncutil/singleflight/singleflight.go:121 +0x2f1
I220525 12:35:49.000498 1 (gostd) testmain.go:78  [-] 1  Test //pkg/sql/opt/exec/execbuilder:execbuilder_test exited with error code 2
```

--- a/pkg/kv/kvserver/liveness/liveness.go
+++ b/pkg/kv/kvserver/liveness/liveness.go
@@ -184,6 +184,7 @@ type HeartbeatCallback func(context.Context)
 // TODO(bdarnell): Also document interaction with draining and decommissioning.
 type NodeLiveness struct {
 	ambientCtx        log.AmbientContext
+	stopper           *stop.Stopper
 	clock             *hlc.Clock
 	db                *kv.DB
 	gossip            *gossip.Gossip
@@ -265,6 +266,7 @@ type Record struct {
 // be moved back and forth.
 type NodeLivenessOptions struct {
 	AmbientCtx              log.AmbientContext
+	Stopper                 *stop.Stopper
 	Settings                *cluster.Settings
 	Gossip                  *gossip.Gossip
 	Clock                   *hlc.Clock
@@ -284,6 +286,7 @@ type NodeLivenessOptions struct {
 func NewNodeLiveness(opts NodeLivenessOptions) *NodeLiveness {
 	nl := &NodeLiveness{
 		ambientCtx:           opts.AmbientCtx,
+		stopper:              opts.Stopper,
 		clock:                opts.Clock,
 		db:                   opts.DB,
 		gossip:               opts.Gossip,
@@ -695,7 +698,6 @@ func (nl *NodeLiveness) IsAvailableNotDraining(nodeID roachpb.NodeID) bool {
 
 // NodeLivenessStartOptions are the arguments to `NodeLiveness.Start`.
 type NodeLivenessStartOptions struct {
-	Stopper *stop.Stopper
 	Engines []storage.Engine
 	// OnSelfLive is invoked after every successful heartbeat
 	// of the local liveness instance's heartbeat loop.
@@ -710,7 +712,7 @@ type NodeLivenessStartOptions struct {
 func (nl *NodeLiveness) Start(ctx context.Context, opts NodeLivenessStartOptions) {
 	log.VEventf(ctx, 1, "starting node liveness instance")
 	retryOpts := base.DefaultRetryOptions()
-	retryOpts.Closer = opts.Stopper.ShouldQuiesce()
+	retryOpts.Closer = nl.stopper.ShouldQuiesce()
 
 	if len(opts.Engines) == 0 {
 		// Avoid silently forgetting to pass the engines. It happened before.
@@ -722,10 +724,10 @@ func (nl *NodeLiveness) Start(ctx context.Context, opts NodeLivenessStartOptions
 	nl.mu.engines = opts.Engines
 	nl.mu.Unlock()
 
-	_ = opts.Stopper.RunAsyncTaskEx(ctx, stop.TaskOpts{TaskName: "liveness-hb", SpanOpt: stop.SterileRootSpan}, func(context.Context) {
+	_ = nl.stopper.RunAsyncTaskEx(ctx, stop.TaskOpts{TaskName: "liveness-hb", SpanOpt: stop.SterileRootSpan}, func(context.Context) {
 		ambient := nl.ambientCtx
 		ambient.AddLogTag("liveness-hb", nil)
-		ctx, cancel := opts.Stopper.WithCancelOnQuiesce(context.Background())
+		ctx, cancel := nl.stopper.WithCancelOnQuiesce(context.Background())
 		defer cancel()
 		ctx, sp := ambient.AnnotateCtxWithSpan(ctx, "liveness heartbeat loop")
 		defer sp.Finish()
@@ -737,7 +739,7 @@ func (nl *NodeLiveness) Start(ctx context.Context, opts NodeLivenessStartOptions
 		for {
 			select {
 			case <-nl.heartbeatToken:
-			case <-opts.Stopper.ShouldQuiesce():
+			case <-nl.stopper.ShouldQuiesce():
 				return
 			}
 			// Give the context a timeout approximately as long as the time we
@@ -777,7 +779,7 @@ func (nl *NodeLiveness) Start(ctx context.Context, opts NodeLivenessStartOptions
 			nl.heartbeatToken <- struct{}{}
 			select {
 			case <-ticker.C:
-			case <-opts.Stopper.ShouldQuiesce():
+			case <-nl.stopper.ShouldQuiesce():
 				return
 			}
 		}

--- a/pkg/kv/kvserver/liveness/liveness.go
+++ b/pkg/kv/kvserver/liveness/liveness.go
@@ -1273,7 +1273,10 @@ func (nl *NodeLiveness) updateLiveness(
 		for i, eng := range engines {
 			eng := eng // pin the loop variable
 			resultCs[i], _ = nl.engineSyncs.DoChan(strconv.Itoa(i), func() (interface{}, error) {
-				return nil, storage.WriteSyncNoop(eng)
+				return nil, nl.stopper.RunTaskWithErr(ctx, "liveness-hb-diskwrite",
+					func(ctx context.Context) error {
+						return storage.WriteSyncNoop(eng)
+					})
 			})
 		}
 		for _, resultC := range resultCs {

--- a/pkg/server/server.go
+++ b/pkg/server/server.go
@@ -400,6 +400,7 @@ func NewServer(cfg Config, stopper *stop.Stopper) (*Server, error) {
 
 	nodeLiveness := liveness.NewNodeLiveness(liveness.NodeLivenessOptions{
 		AmbientCtx:              cfg.AmbientCtx,
+		Stopper:                 stopper,
 		Clock:                   clock,
 		DB:                      db,
 		Gossip:                  g,
@@ -1440,7 +1441,6 @@ func (s *Server) PreStart(ctx context.Context) error {
 	// store "last up" timestamp for every store whenever the liveness record is
 	// updated.
 	s.nodeLiveness.Start(ctx, liveness.NodeLivenessStartOptions{
-		Stopper: s.stopper,
 		Engines: s.engines,
 		OnSelfLive: func(ctx context.Context) {
 			now := s.clock.Now()

--- a/pkg/testutils/localtestcluster/local_test_cluster.go
+++ b/pkg/testutils/localtestcluster/local_test_cluster.go
@@ -184,6 +184,7 @@ func (ltc *LocalTestCluster) Start(t testing.TB, baseCtx *base.Config, initFacto
 	active, renewal := cfg.NodeLivenessDurations()
 	cfg.NodeLiveness = liveness.NewNodeLiveness(liveness.NodeLivenessOptions{
 		AmbientCtx:              cfg.AmbientCtx,
+		Stopper:                 ltc.stopper,
 		Clock:                   cfg.Clock,
 		DB:                      cfg.DB,
 		Gossip:                  cfg.Gossip,
@@ -272,7 +273,7 @@ func (ltc *LocalTestCluster) Start(t testing.TB, baseCtx *base.Config, initFacto
 
 	if !ltc.DisableLivenessHeartbeat {
 		cfg.NodeLiveness.Start(ctx,
-			liveness.NodeLivenessStartOptions{Stopper: ltc.stopper, Engines: []storage.Engine{ltc.Eng}})
+			liveness.NodeLivenessStartOptions{Engines: []storage.Engine{ltc.Eng}})
 	}
 
 	if err := ltc.Store.Start(ctx, ltc.stopper); err != nil {

